### PR TITLE
[Library definitions] Add optional `propTypes` to `React.AbstractComponentStatics`

### LINKED ISSUE #7484: propTypes no longer part of React.AbstractComponentStatics
One cannot reference `propTypes` on a `ComponentType<*>` since 2e7d5a26babe69f2b17cba25f6f62a524534c635.
Is this intended? If yes, do you have a suggestion for handling this?

Example code:
```
import * as React from 'react';

function testHoc<-Props>(Component: React.ComponentType<Props>, props: Props) {
	return class Wrapper extends React.Component<Props> {
      static propTypes = Component.propTypes; // Error here: property `propTypes` is missing in 
`React.AbstractComponentStatics`
      
      render() {
      	return <Component {...props} />
      }
    };
}
```

[Flow try](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoAlgWwA5wE4AuyYAhgM5gBKApiQMZFR5yZgDketDbA3KqlACuAOwbo4wsAWpkCACTh0APAFoACs2xkAfAAoAwi1zDqwggC4qXAgDpDOCaYIAVAJ7ZqSjXC3aANGDYmmSW3loAlGAA3qgAkJwEgniSdDDkFADqeCTYHnhg1AAe0sIAJhQ09Lb2xk5ewdrRqGAtrbIkBOh0gZpuHhQAvGA1jmY2QT59Mnytrc2zYJxl1Hi6kTELYPHUiclgSiMmZtE2pxNaAL5gwNrzsxd3YBd8F0A)

### PR DESCRIPTION
**Pull request by cprodescu**
- fixes #7484
- add test accessing `Component.propTypes`
- fix line numbers for other tests due to shift in `lib/react.js`

### CODE REVIEW

File: lib/react.js [hunk 1]
```diff
@@ -134,6 +134,7 @@ declare type React$AbstractComponentStatics = {
   // This is only on function components, but trying to access name when
   // displayName is undefined is a common pattern.
   name?: ?string,
+  propTypes?: any,
```
Comment (TrySound) [inline lib/react.js]: @jbrown215 This will not let to enable dynamic-export lint ever

Comment (jbrown215) [reply lib/react.js]: Thanks for bringing that up. I am discussing with @dsainati1 and we'll figure out a way to move forward.

Comment (cprodescu) [reply lib/react.js]: Thank you for following up on this :)

**Comment by cprodescu** [comment]
Assuming this can't merge as is due to dynamic-export concerns, I see a few options:
- drop this completely (suggestion on how to handle access to `propTypes` of an abstract component without $FlowIgnore would be helpful)
- try to type the propTypes fully (however this will need to stay in sync with `facebook/prop-types`)
- use a fairly abstract type like `propTypes: {}` (might also be banned by dynamic-export lint later)

Do you see any of these as viable?

**Comment by jbrown215** [comment]
> propTypes: {}

This sounds like a reasonable way to move forward, assuming it works (I don't see why it wouldn't at first glance). Though, I would make it `propTypes?: {}`.

**Comment by jbrown215** [comment]
@cprodescu: Were you able to satisfy your use case with this definition? Would you mind adding a test to tests/react_abstract_component with your use case?

**Comment by jbrown215** [comment]
Also, I can handle re-recording the tests for you.

**Comment by cprodescu** [comment]
I had added the use-case to `tests/react/hoc.js` (since in our xp that's where it occurs), but will move it to `tests/react_abstract_component`. Also still need to rebase.
Thank you for the prompt reply!

**Comment by jbrown215** [comment]
@cprodescu: Class components already allow propTypes, so a test that sets it directly on an AbstractComponent is what we need. Something like:
```
declare var x: React$AbstractComponent<{}, mixed>;
x.propTypes = {foo: 'something'};
```

**Pull request closed without merging**